ann_benchmarks/algorithms/luceneknn.py

```diff
@@ -0,0 +1,124 @@
+        self.M = M
+        self.efConstruction = efConstruction
+
+    def getKnnVectorsFormatForField(self):
+        Lucene94HnswVectorsFormat(self.M, self.efConstruction)
+
+
+class PyLuceneKNN(BaseANN):
+    """
+    KNN using the Lucene Vector datatype.
+    """
+
+    def __init__(self, metric: str, dimension: int, param):
+        try:
+            lucene.initVM(vmargs=['-Djava.awt.headless=true -Xmx6g -Xms6g'])
+        except ValueError:
+            print(f'VM already initialized')
+        self.metric = metric
+        self.dimension = dimension
+        self.param = param
+        self.short_name = f"luceneknn-{param['M']}-{param['efConstruction']}"
+        self.simFunc = VectorSimilarityFunction.DOT_PRODUCT if self.metric == "angular" \
+            else VectorSimilarityFunction.EUCLIDEAN
+        if self.metric not in ("euclidean", "angular"):
+            raise NotImplementedError(f"Not implemented for metric {self.metric}")
+
+    def done(self):
+        if self.dir:
+            self.dir.close()
+
+    def fit(self, X):
+        if self.dimension != X.shape[1]:
+            raise Exception(f"Configured dimension {self.dimension} but data has shape {X.shape}")
+        if self.metric == 'angular':
+            X = sklearn.preprocessing.normalize(X, axis=1, norm='l2')
+        iwc = IndexWriterConfig().setOpenMode(IndexWriterConfig.OpenMode.CREATE)
+        codec = Codec(self.param['M'], self.param['efConstruction'])
+        iwc.setCodec(codec)
+        iwc.setRAMBufferSizeMB(1994.0)
+        self.dir = FSDirectory.open(Paths.get(self.short_name + ".index"))
+        iw = IndexWriter(self.dir, iwc)
+        fieldType = KnnVectorField.createFieldType(self.dimension, self.simFunc)
+        id = 0
+        # X is a numpy matrix, JArray casting only works on python lists.
+        X = X.tolist()
+        for x in X:
+            doc = Document()
+            doc.add(KnnVectorField("knn", JArray('float')(x), fieldType))
+            doc.add(StoredField("id", id))
+            iw.addDocument(doc)
+            id += 1
+            if id + 1 % 1000 == 0:
+                print(f"LuceneKNN: written {id} docs")
+        # Force merge so only one HNSW graph is searched.
+        iw.forceMerge(1)
+        print(f"LuceneKNN: written {id} docs")
+        iw.close()
+        self.searcher = IndexSearcher(DirectoryReader.open(self.dir))
+
+    def set_query_arguments(self, ef):
+        self.name = f"luceneknn-{self.dimension}-{self.param['M']}-{self.param['efConstruction']}-{ef}"
+        self.ef = ef
+
+    def query(self, q, n):
+        if self.metric == 'angular':
+            q = q / np.linalg.norm(q)
+        return self.run_knn_query(num_candidates=self.ef, n=n, q=q.tolist())
+
+    def prepare_batch_query(self, X, n):
+        if self.metric == 'angular':
+            X = sklearn.preprocessing.normalize(X, axis=1, norm='l2')
+        self.queries = X.tolist()
+        self.topK = n
+
+    def run_batch_query(self):
```

Comment:
What's the difference between `run_batch_query` and `batch_query`?

Reply:
See: https://github.com/erikbern/ann-benchmarks/blob/6ad14dc08ea2fa572df7339a9f024975bac1a309/ann_benchmarks/runner.py#L54

One runs the preparation, the other doesn't. Honestly, I don't know when the runner chooses one path or the other, but satisfied both as it isn't difficult.

Reply:
hmm, it's just a bit confusing since one implementation is parallel and the other is serial.